questions/250/explanation.md

```diff
@@ -1,62 +1,62 @@
+
+We start by deducing the template arguments for both overloads. For `foo₁`, the argument `1` is used to deduce `T` to be `int`. It's signature is then `foo₁(int, ...)`. However, for overload resolution we ignore the `...` since it doesn't have a matching argument: §[over.match.viable]¶2.3:
+
+> For the purposes of overload resolution, the parameter list is truncated on the right, so that there are exactly *m* parameters.
+
+So we end up with the viable function `foo₁(int)`.
+
+For `foo₂`, `T...` is a function parameter pack. The argument `1`  is used to deduce this as one `int`, and it's signature is `foo₂(int)`.
+
+So, both `foo`s are viable. Which one is best? §[over.match.best]¶2:
+> Given these definitions, a viable function `F1` is defined to be a *better* function than another viable function `F2` if for all arguments *i*, `ICSi(F1)` is not a worse conversion sequence than `ICSi(F2)`, and then
+>
+>for some argument *j*, `ICSj(F1)` is a better conversion sequence than `ICSj(F2)`
+
+`ICSi(F1)` just means "the **I**mplicit **C**onversion **S**equence for argument ***i***. Our sole argument `1` is of type `int`, so no conversion sequence is needed for neither  `foo₁` nor `foo₂`, as they both take an `int` as their first argument. (i.e. neither is a better conversion sequence than the other.)  
+
+§[over.match.best]¶2.5 continues:
+> [or, if not that] F1 and F2 are function template specializations, and the function template for F1 is more specialized than the template for F2 according to the partial ordering rules
+
+I won't go into all the details for partial ordering here, but skip to the interesting parts. In short, we transform each specialization by substituting each template parameter with a unique, made up type and get  `foo₁(X, ...)` and `foo₂(Y)`. Then we do deduction from each transformed function to the other, original template.
+
+Again, the ellipsis is ignored since it doesn't have an argument at the call site: §[temp.deduct.partial]¶3:
+
+> The types used to determine the ordering depend on the context in which the partial ordering is done:
+> - In the context of a function call, the types used are those function parameter types for which the function call has arguments.
+
+We now do deduction of  `foo₂(T...)` from `foo₁(X)`, which deduces `T=X`. We then do deduction of `foo₁(T)` from `foo₂(Y)`. It looks like we would get `T=Y`, but §[temp.deduct.type]¶10 says:
+
+> During partial ordering, if `Ai` was originally a function parameter pack [and] `Pi` is not a function parameter pack, template argument deduction fails.
+
+`Ai` in this case is `Y`, which was originally a function parameter pack. `Pi` is `T`, which is not a function parameter pack, so deduction fails.
+
+Since we can deduce `foo₂` from `foo₁`, but not `foo₁` from `foo₂`,  `foo₁` is more specialized and a better function. Overload resolution picks it and prints `A`.
```

Comment:
nit:

```suggestion
`A`<sub>*i*<sub> in this case is `Y`, which was originally a function parameter pack. `P`<sub>*i*<sub> is `T`, which is not a function parameter pack, so deduction fails.

Since we can deduce `foo₂` from `foo₁`, but not `foo₁` from `foo₂`, `foo₁` is more specialized and a better function. Overload resolution picks it and prints `A`.
```